distributed/protocol/serialize.py

```diff
@@ -135,6 +135,48 @@ def serialize(x, serializers=None, on_error="message", context=None):
     if isinstance(x, Serialized):
         return x.header, x.frames
 
+    if type(x) in (dict, list, set, tuple) and len(x) <= 5:
+        if isinstance(x, dict):
+            headers_frames = []
+            for k, v in x.items():
+                _header, _frames = serialize(
+                    v, serializers=serializers, on_error=on_error, context=context
+                )
+                try:
+                    k = msgpack.dumps(k)
+                except Exception:
+                    k = serialize(
+                        k, serializers=serializers, on_error=on_error, context=context
+                    )
+                else:
+                    _header["key"] = k
```

Comment:
There are probably situations where the frames produced by serializing `k` aren't msgpack encodable.  Also, if `k` is large then we wouldn't want it in the header, but would prefer to pass on this collections optimization for serialization.  This is likely to be a rare case, it's odd to see people place large complex values in keys, but it's still something that we should avoid.

```python
if type(obj) is dict and len(obj) <=5 :
    try:
        msgpack.dumps(list(obj.keys()))
    except Exception:
        dict_safe = False
    else:
        dict_safe = True

if type(obj) is dict and dict_safe or type(obj) in (tuple, list, set):
    ... # the rest of your code
```

Reply:
Done.